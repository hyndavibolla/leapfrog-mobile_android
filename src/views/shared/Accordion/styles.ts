import { css } from '@emotion/native';

import { FONT_SIZE, COLOR } from '../../../constants';

export const styles = {
  container: css`
    flex-direction: column;
    width: 100%;
  `,
  paddingDefaultContainer: css`
    padding-horizontal: 8px;
  `,
  titleContainer: css`
    flex-direction: row;
    align-items: flex-start;
    justify-content: space-between;
  `,
  textContainer: css`
    margin-top: 16px;
    flex-direction: row;
    width: 100%;
    overflow: hidden;
  `,
  text: css`
    font-size: ${FONT_SIZE.SMALL};
    color: ${COLOR.DARK_GRAY};
    line-height: 28px;
    text-align: left;
  `,
  hidden: css`
    height: 0;
    margin-top: 0;
  `
};
