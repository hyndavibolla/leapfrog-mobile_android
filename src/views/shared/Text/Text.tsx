import React, { memo, ReactNode } from 'react';
import { TextProps } from 'react-native';

import { COLOR, FONT_FAMILY, FONT_SIZE, LINE_HEIGHT } from '_constants';

import { StyledText } from './styles';

export interface Props extends TextProps {
  color?: COLOR;
  font?: FONT_FAMILY;
  size?: FONT_SIZE;
  lineHeight?: LINE_HEIGHT;
  children: ReactNode;
}

const Text = ({ color, font, size, lineHeight, children, ...otherProps }: Props) => {
  return (
    <StyledText color={color} font={font} size={size} lineHeight={lineHeight} {...otherProps}>
      {children}
    </StyledText>
  );
};

export default memo(Text);
