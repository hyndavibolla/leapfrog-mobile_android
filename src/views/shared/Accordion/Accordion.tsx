import React, { memo, useState, PropsWithChildren, useCallback } from 'react';
import { View, TouchableHighlight, StyleProp, TextStyle, ViewStyle } from 'react-native';
import { View as AnimatedView } from 'react-native-animatable';

import { Icon } from '_commons/components/atoms/Icon';
import { styles } from './styles';
import { Title, TitleType } from '../Title';
import { useTestingHelper } from '_utils/useTestingHelper';
import { Text } from '../Text';
import { COLOR, FONT_SIZE, ICON } from '_constants';

export type Props = PropsWithChildren<{
  title: string;
  startsOpen?: boolean;
  titleStyle?: StyleProp<TextStyle>;
  textStyle?: StyleProp<TextStyle>;
  wrapperStyle?: StyleProp<ViewStyle> | StyleProp<TextStyle>;
  noUseDefaultPadding?: boolean;
}>;

export const Accordion = ({ title, startsOpen = true, noUseDefaultPadding = false, textStyle, titleStyle, wrapperStyle, children }: Props) => {
  const { getTestIdProps } = useTestingHelper('accordion');
  const [isOpen, setIsOpen] = useState(!!startsOpen);
  const Wrapper = ['string', 'number'].includes(typeof children) ? Text : View;
  const onToggle = useCallback(() => setIsOpen(prevState => !prevState), []);
  const duration = 300;
  const rotateDownAnimation = { 0: { rotate: '90deg' }, 1: { rotate: '-90deg' } };
  const rotateUpAnimation = { 0: { rotate: '-90deg' }, 1: { rotate: '90deg' } };
  return (
    <View style={styles.container}>
      <TouchableHighlight onPress={onToggle} {...getTestIdProps('btn')} underlayColor="transparent" activeOpacity={1}>
        <View style={[styles.titleContainer, !noUseDefaultPadding && styles.paddingDefaultContainer]}>
          <Title type={TitleType.SECTION} style={titleStyle}>
            {title}
          </Title>
          <AnimatedView animation={(isOpen ? rotateDownAnimation : rotateUpAnimation) as any} duration={duration}>
            <Icon name={ICON.BRACKET_RIGHT} size={FONT_SIZE.REGULAR} color={COLOR.DARK_GRAY} />
          </AnimatedView>
        </View>
      </TouchableHighlight>
      <View style={[styles.textContainer, !noUseDefaultPadding && styles.paddingDefaultContainer, !isOpen && styles.hidden]}>
        <Wrapper style={[styles.text, textStyle, wrapperStyle]}>{children}</Wrapper>
      </View>
    </View>
  );
};

export default memo(Accordion);
