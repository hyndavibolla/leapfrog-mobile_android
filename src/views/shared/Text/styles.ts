import styled from '@emotion/native';

import { FONT_FAMILY, COLOR, FONT_SIZE, LINE_HEIGHT } from '_constants';

import { Props } from './Text';

export const StyledText = styled.Text<Props>`
  font-family: ${({ font }: Props) => font || FONT_FAMILY.MEDIUM};
  color: ${({ color }: Props) => color || COLOR.DARK_GRAY};
  font-size: ${({ size }: Props) => size || FONT_SIZE.SMALLER};
  line-height: ${({ lineHeight }: Props) => lineHeight || LINE_HEIGHT.PARAGRAPH};
`;
