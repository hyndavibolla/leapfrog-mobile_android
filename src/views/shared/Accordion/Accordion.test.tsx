import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { View } from 'react-native';

import Accordion, { Props } from './Accordion';
import { Text } from '../Text';

describe('Accordion', () => {
  let props: Props;
  beforeEach(() => {
    props = { title: 'Title', children: 'Accordion text' };
  });

  it('should render', () => {
    const { toJSON } = render(<Accordion {...props} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render with complex content', () => {
    const { toJSON } = render(
      <Accordion {...props}>
        <View>
          <Text>i am a complex content</Text>
        </View>
      </Accordion>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render starting collapses', () => {
    const { toJSON } = render(<Accordion {...props} startsOpen={false} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should show and hide accordion content', () => {
    const { getByTestId, toJSON } = render(<Accordion {...props} />);
    fireEvent.press(getByTestId('accordion-btn')); // close accordion content
    expect(toJSON()).toMatchSnapshot();
    fireEvent.press(getByTestId('accordion-btn')); // open it
    expect(toJSON()).toMatchSnapshot();
  });
});
