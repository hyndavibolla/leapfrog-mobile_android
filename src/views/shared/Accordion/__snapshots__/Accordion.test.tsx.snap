// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion should render 1`] = `
<View
  style={
    Object {
      "flexDirection": "column",
      "width": "100%",
    }
  }
>
  <View
    accessibilityLabel="accordion-btn"
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    testID="accordion-btn"
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "flex-start",
            "flexDirection": "row",
            "justifyContent": "space-between",
          },
          Object {
            "paddingHorizontal": 8,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#262C2F",
              "fontFamily": "SFProDisplay-Bold",
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
          ]
        }
        type="sectionHeader"
      >
        Title
      </Text>
      <div
        animation={
          Object {
            "0": Object {
              "rotate": "90deg",
            },
            "1": Object {
              "rotate": "-90deg",
            },
          }
        }
        duration={300}
      >
        <Text
          accessibilityLabel="icon-text"
          accessible={true}
          allowFontScaling={false}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#909BA0",
                "fontSize": 20,
              },
              undefined,
              Object {
                "fontFamily": "sywmax",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
          testID="icon-text"
        >
          
        </Text>
      </div>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "marginTop": 16,
          "overflow": "hidden",
          "width": "100%",
        },
        Object {
          "paddingHorizontal": 8,
        },
        false,
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#909BA0",
            "fontFamily": "SFProDisplay-Medium",
            "fontSize": 16,
            "lineHeight": 22,
          },
          Array [
            Object {
              "color": "#909BA0",
              "fontSize": 18,
              "lineHeight": 28,
              "textAlign": "left",
            },
            undefined,
            undefined,
          ],
        ]
      }
    >
      Accordion text
    </Text>
  </View>
</View>
`;

exports[`Accordion should render starting collapses 1`] = `
<View
  style={
    Object {
      "flexDirection": "column",
      "width": "100%",
    }
  }
>
  <View
    accessibilityLabel="accordion-btn"
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    testID="accordion-btn"
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "flex-start",
            "flexDirection": "row",
            "justifyContent": "space-between",
          },
          Object {
            "paddingHorizontal": 8,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#262C2F",
              "fontFamily": "SFProDisplay-Bold",
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
          ]
        }
        type="sectionHeader"
      >
        Title
      </Text>
      <div
        animation={
          Object {
            "0": Object {
              "rotate": "-90deg",
            },
            "1": Object {
              "rotate": "90deg",
            },
          }
        }
        duration={300}
      >
        <Text
          accessibilityLabel="icon-text"
          accessible={true}
          allowFontScaling={false}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#909BA0",
                "fontSize": 20,
              },
              undefined,
              Object {
                "fontFamily": "sywmax",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
          testID="icon-text"
        >
          
        </Text>
      </div>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "marginTop": 16,
          "overflow": "hidden",
          "width": "100%",
        },
        Object {
          "paddingHorizontal": 8,
        },
        Object {
          "height": 0,
          "marginTop": 0,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#909BA0",
            "fontFamily": "SFProDisplay-Medium",
            "fontSize": 16,
            "lineHeight": 22,
          },
          Array [
            Object {
              "color": "#909BA0",
              "fontSize": 18,
              "lineHeight": 28,
              "textAlign": "left",
            },
            undefined,
            undefined,
          ],
        ]
      }
    >
      Accordion text
    </Text>
  </View>
</View>
`;

exports[`Accordion should render with complex content 1`] = `
<View
  style={
    Object {
      "flexDirection": "column",
      "width": "100%",
    }
  }
>
  <View
    accessibilityLabel="accordion-btn"
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    testID="accordion-btn"
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "flex-start",
            "flexDirection": "row",
            "justifyContent": "space-between",
          },
          Object {
            "paddingHorizontal": 8,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#262C2F",
              "fontFamily": "SFProDisplay-Bold",
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
          ]
        }
        type="sectionHeader"
      >
        Title
      </Text>
      <div
        animation={
          Object {
            "0": Object {
              "rotate": "90deg",
            },
            "1": Object {
              "rotate": "-90deg",
            },
          }
        }
        duration={300}
      >
        <Text
          accessibilityLabel="icon-text"
          accessible={true}
          allowFontScaling={false}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#909BA0",
                "fontSize": 20,
              },
              undefined,
              Object {
                "fontFamily": "sywmax",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
          testID="icon-text"
        >
          
        </Text>
      </div>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "marginTop": 16,
          "overflow": "hidden",
          "width": "100%",
        },
        Object {
          "paddingHorizontal": 8,
        },
        false,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "color": "#909BA0",
            "fontSize": 18,
            "lineHeight": 28,
            "textAlign": "left",
          },
          undefined,
          undefined,
        ]
      }
    >
      <View>
        <Text
          style={
            Array [
              Object {
                "color": "#909BA0",
                "fontFamily": "SFProDisplay-Medium",
                "fontSize": 16,
                "lineHeight": 22,
              },
              undefined,
            ]
          }
        >
          i am a complex content
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`Accordion should show and hide accordion content 1`] = `
<View
  style={
    Object {
      "flexDirection": "column",
      "width": "100%",
    }
  }
>
  <View
    accessibilityLabel="accordion-btn"
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    testID="accordion-btn"
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "flex-start",
            "flexDirection": "row",
            "justifyContent": "space-between",
          },
          Object {
            "paddingHorizontal": 8,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#262C2F",
              "fontFamily": "SFProDisplay-Bold",
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
          ]
        }
        type="sectionHeader"
      >
        Title
      </Text>
      <div
        animation={
          Object {
            "0": Object {
              "rotate": "-90deg",
            },
            "1": Object {
              "rotate": "90deg",
            },
          }
        }
        duration={300}
      >
        <Text
          accessibilityLabel="icon-text"
          accessible={true}
          allowFontScaling={false}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#909BA0",
                "fontSize": 20,
              },
              undefined,
              Object {
                "fontFamily": "sywmax",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
          testID="icon-text"
        >
          
        </Text>
      </div>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "marginTop": 16,
          "overflow": "hidden",
          "width": "100%",
        },
        Object {
          "paddingHorizontal": 8,
        },
        Object {
          "height": 0,
          "marginTop": 0,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#909BA0",
            "fontFamily": "SFProDisplay-Medium",
            "fontSize": 16,
            "lineHeight": 22,
          },
          Array [
            Object {
              "color": "#909BA0",
              "fontSize": 18,
              "lineHeight": 28,
              "textAlign": "left",
            },
            undefined,
            undefined,
          ],
        ]
      }
    >
      Accordion text
    </Text>
  </View>
</View>
`;

exports[`Accordion should show and hide accordion content 2`] = `
<View
  style={
    Object {
      "flexDirection": "column",
      "width": "100%",
    }
  }
>
  <View
    accessibilityLabel="accordion-btn"
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    testID="accordion-btn"
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "flex-start",
            "flexDirection": "row",
            "justifyContent": "space-between",
          },
          Object {
            "paddingHorizontal": 8,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#262C2F",
              "fontFamily": "SFProDisplay-Bold",
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
          ]
        }
        type="sectionHeader"
      >
        Title
      </Text>
      <div
        animation={
          Object {
            "0": Object {
              "rotate": "90deg",
            },
            "1": Object {
              "rotate": "-90deg",
            },
          }
        }
        duration={300}
      >
        <Text
          accessibilityLabel="icon-text"
          accessible={true}
          allowFontScaling={false}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#909BA0",
                "fontSize": 20,
              },
              undefined,
              Object {
                "fontFamily": "sywmax",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
          testID="icon-text"
        >
          
        </Text>
      </div>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "marginTop": 16,
          "overflow": "hidden",
          "width": "100%",
        },
        Object {
          "paddingHorizontal": 8,
        },
        false,
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#909BA0",
            "fontFamily": "SFProDisplay-Medium",
            "fontSize": 16,
            "lineHeight": 22,
          },
          Array [
            Object {
              "color": "#909BA0",
              "fontSize": 18,
              "lineHeight": 28,
              "textAlign": "left",
            },
            undefined,
            undefined,
          ],
        ]
      }
    >
      Accordion text
    </Text>
  </View>
</View>
`;
